File: test/_harnesses/go/basic/basic.go
package main

import (
	"flag"
	"io"
	"os"
	"path/filepath"

	"messages/comprehensive_message"
)

var ok bool = true

func softAssert(condition bool, label string) {
	if !condition {
		os.Stderr.WriteString("FAILED! Go: " + label + "\n")
		ok = false
	}
}

func main() {
	var example comprehensive_message.TestingMessage
	example.B = 250
	example.Tf = true
	example.I16 = -32000
	example.Ui16 = 65000
	example.I32 = -2000000000
	example.Ui32 = 4000000000
	example.I64 = -9000000000000000000
	example.Ui64 = 18000000000000000000
	example.F = 3.1415927410125732421875
	example.D = 2.718281828459045090795598298427648842334747314453125
	example.Ee = comprehensive_message.EnumeratedBeta
	example.Es = comprehensive_message.SpecifiedNegative
	example.S = "Lorem ipsum dolor sit amet, consectetur adipiscing elit."
	example.V2.X = 256.512
	example.V2.Y = 1024.768
	example.V3.X = 128.64
	example.V3.Y = 2048.4096
	example.V3.Z = 16.32
	example.C.R = 255
	example.C.G = 128
	example.C.B = 0
	example.Il = []int16{-1000, 500, 0, 750, 2000}
	example.El = []comprehensive_message.Specified{
		comprehensive_message.SpecifiedNegative,
		comprehensive_message.SpecifiedNegative,
		comprehensive_message.SpecifiedPositive,
		comprehensive_message.SpecifiedZero,
		comprehensive_message.SpecifiedPositive,
		comprehensive_message.SpecifiedZero,
	}
	example.Sl = []string{
		"Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
		"Quisque est eros, placerat ut libero ut, pellentesque tincidunt sem.",
		"Vivamus pellentesque turpis aliquet pretium tincidunt.",
		"Nulla facilisi.",
		"🐼❤️✝️",
		"用ねぼ雪入文モ段足リフケ報通ンさーを応細めい気川ヤセ車不古6治ニフサコ悩段をご青止ぽっ期年ト量報驚テルユ役1家埋詰軟きぎ。",
		"لآخر نشجب ونستنكر هؤلاء الرجال المفتونون بنشوة اللحظة الهائمون في رغبات",
	}
	var v21 comprehensive_message.Vec2
	v21.X = 10.0
	v21.Y = 15.0
	var v22 comprehensive_message.Vec2
	v22.X = 20.0
	v22.Y = 25.0
	var v23 comprehensive_message.Vec2
	v23.X = 30.0
	v23.Y = 35.0
	var v24 comprehensive_message.Vec2
	v24.X = 40.0
	v24.Y = 45.0
	example.V2l = []comprehensive_message.Vec2{v21, v22, v23, v24}
	var v31 comprehensive_message.Vec3
	v31.X = 10.0
	v31.Y = 15.0
	v31.Z = 17.5
	var v32 comprehensive_message.Vec3
	v32.X = 20.0
	v32.Y = 25.0
	v32.Z = 27.5
	var v33 comprehensive_message.Vec3
	v33.X = 30.0
	v33.Y = 35.0
	v33.Z = 37.5
	var v34 comprehensive_message.Vec3
	v34.X = 40.0
	v34.Y = 45.0
	v34.Z = 47.5
	example.V3l = []comprehensive_message.Vec3{v31, v32, v33, v34}
	var c1 comprehensive_message.Color
	c1.R = 255
	c1.G = 0
	c1.B = 0
	var c2 comprehensive_message.Color
	c2.R = 0
	c2.G = 255
	c2.B = 0
	var c3 comprehensive_message.Color
	c3.R = 0
	c3.G = 0
	c3.B = 255
	example.Cl = []comprehensive_message.Color{c1, c2, c3}
	example.Cx.Identifier = 127
	example.Cx.Label = "ComplexDataObject"
	example.Cx.BackgroundColor = c1
	example.Cx.TextColor = c2
	example.Cx.Spectrum = []comprehensive_message.Color{c3, c2, c1}
	example.Cx.Ranges = []comprehensive_message.Specified{
		comprehensive_message.SpecifiedNegative,
		comprehensive_message.SpecifiedPositive,
	}
	var cx1 comprehensive_message.ComplexData
	cx1.Identifier = 255
	cx1.Label = "Complex1"
	cx1.BackgroundColor = c3
	cx1.TextColor = c1
	cx1.Spectrum = []comprehensive_message.Color{c3, c2, c1, c2, c3}
	cx1.Ranges = []comprehensive_message.Specified{
		comprehensive_message.SpecifiedZero,
		comprehensive_message.SpecifiedPositive,
	}
	var cx2 comprehensive_message.ComplexData
	cx2.Identifier = 63
	cx2.Label = "Complex2"
	cx2.BackgroundColor = c1
	cx2.TextColor = c3
	cx2.Spectrum = []comprehensive_message.Color{c1, c2, c3, c2, c1}
	cx2.Ranges = []comprehensive_message.Specified{
		comprehensive_message.SpecifiedNegative,
		comprehensive_message.SpecifiedZero,
	}
	example.Cxl = []comprehensive_message.ComplexData{cx1, cx2}

	readPathPtr := flag.String("read", "", "path to message file for verification")
	generatePathPtr := flag.String("generate", "", "path to message file for generation")
	flag.Parse()

	if len(*generatePathPtr) > 0 {
		os.MkdirAll(filepath.Dir(*generatePathPtr), os.ModePerm)
		dat, err := os.Create(*generatePathPtr)
		if err != nil {
			panic(err)
		}
		defer dat.Close()

		example.WriteBytes(dat, false)

		softAssert(example.GetSizeInBytes() == 956, "size calculation check")
		seek, _ := dat.Seek(0, io.SeekCurrent)
		softAssert(example.GetSizeInBytes() == (int)(seek), "written bytes check")
	} else if len(*readPathPtr) > 0 {
		dat, err := os.Open(*readPathPtr)
		if err != nil {
			panic(err)
		}
		defer dat.Close()

		input, err := comprehensive_message.TestingMessageFromBytes(dat)
		softAssert(err == nil, "parsing test message")

		softAssert(input.B == example.B, "byte")
		softAssert(input.Tf == example.Tf, "bool")
		softAssert(input.I16 == example.I16, "i16")
		softAssert(input.Ui16 == example.Ui16, "ui16")
		softAssert(input.I32 == example.I32, "i32")
		softAssert(input.Ui32 == example.Ui32, "ui32")
		softAssert(input.I64 == example.I64, "i64")
		softAssert(input.Ui64 == example.Ui64, "ui64")
		softAssert(input.F == example.F, "float")
		softAssert(input.D == example.D, "double")
		softAssert(input.Ee == example.Ee, "enumerated")
		softAssert(input.Es == example.Es, "specified")
		softAssert(input.S == example.S, "string")
		softAssert(input.V2.X == example.V2.X, "Vec2")
		softAssert(input.V2.Y == example.V2.Y, "Vec2")
		softAssert(input.V3.X == example.V3.X, "Vec3")
		softAssert(input.V3.Y == example.V3.Y, "Vec3")
		softAssert(input.V3.Z == example.V3.Z, "Vec3")
		softAssert(input.C.R == example.C.R, "Color")
		softAssert(input.C.G == example.C.G, "Color")
		softAssert(input.C.B == example.C.B, "Color")
		softAssert(len(input.Il) == len(example.Il), "[int16].length")
		for i := 0; i < len(input.Il); i++ {
			softAssert(input.Il[i] == example.Il[i], "[int16]")
		}
		softAssert((len(input.El) == len(example.El)), "[Specified].length")
		for i := 0; i < len(input.El); i++ {
			softAssert(input.El[i] == example.El[i], "[Specified]")
		}
		softAssert(len(input.Sl) == len(example.Sl), "[string].length")
		for i := 0; i < len(input.Sl); i++ {
			softAssert(input.Sl[i] == example.Sl[i], "[string]")
		}
		softAssert(len(input.V2l) == len(example.V2l), "[Vec2].length")
		for i := 0; i < len(input.V2l); i++ {
			softAssert(input.V2l[i].X == example.V2l[i].X, "[Vec2].x")
			softAssert(input.V2l[i].Y == example.V2l[i].Y, "[Vec2].y")
		}
		softAssert(len(input.V3l) == len(example.V3l), "[Vec3].length")
		for i := 0; i < len(input.V3l); i++ {
			softAssert(input.V3l[i].X == example.V3l[i].X, "[Vec3].x")
			softAssert(input.V3l[i].Y == example.V3l[i].Y, "[Vec3].y")
			softAssert(input.V3l[i].Z == example.V3l[i].Z, "[Vec3].z")
		}
		softAssert(len(input.Cl) == len(example.Cl), "[Color].length")
		for i := 0; i < len(input.Cl); i++ {
			softAssert(input.Cl[i].R == example.Cl[i].R, "[Color].r")
			softAssert(input.Cl[i].G == example.Cl[i].G, "[Color].g")
			softAssert(input.Cl[i].B == example.Cl[i].B, "[Color].b")
		}
		softAssert(input.Cx.Identifier == example.Cx.Identifier, "ComplexData.identifier")
		softAssert(input.Cx.Label == example.Cx.Label, "ComplexData.label")
		softAssert(input.Cx.BackgroundColor.R == example.Cx.BackgroundColor.R, "ComplexData.BackgroundColor.r")
		softAssert(input.Cx.BackgroundColor.G == example.Cx.BackgroundColor.G, "ComplexData.BackgroundColor.g")
		softAssert(input.Cx.BackgroundColor.B == example.Cx.BackgroundColor.B, "ComplexData.BackgroundColor.b")
		softAssert(input.Cx.TextColor.R == example.Cx.TextColor.R, "ComplexData.TextColor.r")
		softAssert(input.Cx.TextColor.G == example.Cx.TextColor.G, "ComplexData.TextColor.g")
		softAssert(input.Cx.TextColor.B == example.Cx.TextColor.B, "ComplexData.TextColor.b")
		softAssert(len(input.Cx.Spectrum) == len(example.Cx.Spectrum), "ComplexData.spectrum.length")
		for i := 0; i < len(input.Cx.Spectrum); i++ {
			softAssert(input.Cx.Spectrum[i].R == example.Cx.Spectrum[i].R, "ComplexData.spectrum.r")
			softAssert(input.Cx.Spectrum[i].G == example.Cx.Spectrum[i].G, "ComplexData.spectrum.g")
			softAssert(input.Cx.Spectrum[i].B == example.Cx.Spectrum[i].B, "ComplexData.spectrum.b")
		}
		softAssert(len(input.Cx.Ranges) == len(example.Cx.Ranges), "ComplexData.ranges.length")
		for i := 0; i < len(input.Cx.Ranges); i++ {
			softAssert(input.Cx.Ranges[i] == example.Cx.Ranges[i], "ComplexData.ranges")
		}
		softAssert(len(input.Cxl) == len(example.Cxl), "[ComplexData].length")
		for i := 0; i < len(input.Cxl); i++ {
			softAssert(input.Cxl[i].Identifier == example.Cxl[i].Identifier, "[ComplexData].identifier")
			softAssert(input.Cxl[i].Label == example.Cxl[i].Label, "[ComplexData].label")
			softAssert(input.Cxl[i].BackgroundColor.R == example.Cxl[i].BackgroundColor.R, "[ComplexData].backgroundColor.r")
			softAssert(input.Cxl[i].BackgroundColor.G == example.Cxl[i].BackgroundColor.G, "[ComplexData].backgroundColor.g")
			softAssert(input.Cxl[i].BackgroundColor.B == example.Cxl[i].BackgroundColor.B, "[ComplexData].backgroundColor.b")
			softAssert(input.Cxl[i].TextColor.R == example.Cxl[i].TextColor.R, "[ComplexData].textColor.r")
			softAssert(input.Cxl[i].TextColor.G == example.Cxl[i].TextColor.G, "[ComplexData].textColor.g")
			softAssert(input.Cxl[i].TextColor.B == example.Cxl[i].TextColor.B, "[ComplexData].textColor.b")
			softAssert(len(input.Cxl[i].Ranges) == len(example.Cxl[i].Ranges), "[ComplexData].ranges.length")
			for j := 0; j < len(input.Cxl[i].Ranges); j++ {
				softAssert(input.Cxl[i].Ranges[j] == example.Cxl[i].Ranges[j], "[ComplexData].ranges")
			}
		}

	}

	if !ok {
		os.Stderr.WriteString("Failed assertions.\n")
		os.Exit(1)
	}
}

File: test/_harnesses/assemblyscript/assembly/basic.ts
import * as harness from "./_harness";
export * from "./_harness";

import * as ComprehensiveMessage from '../../../../out/generated/assemblyscript/ComprehensiveMessage';

const example = new ComprehensiveMessage.TestingMessage();
example.b = 250;
example.tf = true;
example.i16 = -32000;
example.ui16 = 65000;
example.i32 = -2000000000;
example.ui32 = 4000000000;
example.i64 = -9000000000000000000;
example.ui64 = 18000000000000000000;
example.f = 3.1415927410125732421875;
example.d = 2.718281828459045090795598298427648842334747314453125;
example.ee = ComprehensiveMessage.Enumerated.Beta;
example.es = ComprehensiveMessage.Specified.Negative;
example.s = "Lorem ipsum dolor sit amet, consectetur adipiscing elit.";
example.v2 = new ComprehensiveMessage.Vec2();
example.v2.x = 256.512;
example.v2.y = 1024.768;
example.v3 = new ComprehensiveMessage.Vec3();
example.v3.x = 128.64;
example.v3.y = 2048.4096;
example.v3.z = 16.32;
example.c = new ComprehensiveMessage.Color();
example.c.r = 255;
example.c.g = 128;
example.c.b = 0;
example.il = [-1000, 500, 0, 750, 2000];
example.el = [
    ComprehensiveMessage.Specified.Negative,
    ComprehensiveMessage.Specified.Negative,
    ComprehensiveMessage.Specified.Positive,
    ComprehensiveMessage.Specified.Zero,
    ComprehensiveMessage.Specified.Positive,
    ComprehensiveMessage.Specified.Zero
],
example.sl = [
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    "Quisque est eros, placerat ut libero ut, pellentesque tincidunt sem.",
    "Vivamus pellentesque turpis aliquet pretium tincidunt.",
    "Nulla facilisi.",
    "🐼❤️✝️",
    "用ねぼ雪入文モ段足リフケ報通ンさーを応細めい気川ヤセ車不古6治ニフサコ悩段をご青止ぽっ期年ト量報驚テルユ役1家埋詰軟きぎ。",
    "لآخر نشجب ونستنكر هؤلاء الرجال المفتونون بنشوة اللحظة الهائمون في رغبات",
];
const v21 = new ComprehensiveMessage.Vec2();
v21.x = 10.0;
v21.y = 15.0;
const v22 = new ComprehensiveMessage.Vec2();
v22.x = 20.0;
v22.y = 25.0;
const v23 = new ComprehensiveMessage.Vec2();
v23.x = 30.0;
v23.y = 35.0;
const v24 = new ComprehensiveMessage.Vec2();
v24.x = 40.0;
v24.y = 45.0;
example.v2l = [
    v21, v22, v23, v24
];
const v31 = new ComprehensiveMessage.Vec3();
v31.x = 10.0;
v31.y = 15.0;
v31.z = 17.5;
const v32 = new ComprehensiveMessage.Vec3();
v32.x = 20.0;
v32.y = 25.0;
v32.z = 27.5;
const v33 = new ComprehensiveMessage.Vec3();
v33.x = 30.0;
v33.y = 35.0;
v33.z = 37.5;
const v34 = new ComprehensiveMessage.Vec3();
v34.x = 40.0;
v34.y = 45.0;
v34.z = 47.5;
example.v3l = [
    v31, v32, v33, v34
];
const c1 = new ComprehensiveMessage.Color();
c1.r = 255;
c1.g = 0;
c1.b = 0;
const c2 = new ComprehensiveMessage.Color();
c2.r = 0;
c2.g = 255;
c2.b = 0;
const c3 = new ComprehensiveMessage.Color();
c3.r = 0;
c3.g = 0;
c3.b = 255;
example.cl = [
    c1, c2, c3
];
example.cx = new ComprehensiveMessage.ComplexData();
example.cx.identifier = 127;
example.cx.label = "ComplexDataObject";
example.cx.backgroundColor = c1;
example.cx.textColor = c2;
example.cx.spectrum = [c3, c2, c1];
example.cx.ranges = [
    ComprehensiveMessage.Specified.Negative,
    ComprehensiveMessage.Specified.Positive,
];
const cx1 = new ComprehensiveMessage.ComplexData();
cx1.identifier = 255;
cx1.label = "Complex1";
cx1.backgroundColor = c3;
cx1.textColor = c1;
cx1.spectrum = [c3, c2, c1, c2, c3];
cx1.ranges = [
    ComprehensiveMessage.Specified.Zero,
    ComprehensiveMessage.Specified.Positive,
];
const cx2 = new ComprehensiveMessage.ComplexData();
cx2.identifier = 63;
cx2.label = "Complex2";
cx2.backgroundColor = c1;
cx2.textColor = c3;
cx2.spectrum = [c1, c2, c3, c2, c1];
cx2.ranges = [
    ComprehensiveMessage.Specified.Negative,
    ComprehensiveMessage.Specified.Zero,
];
example.cxl = [cx1, cx2];

export function generate(): usize {
    const byteLen = example.getSizeInBytes();
    harness.allocate(byteLen);
    const dv = harness.getDataView();
    example.writeBytes(dv, false);

    harness.softAssert(byteLen == 956, "size calculation check");

    return harness.getDataPtr();
}

export function read(): void {
    const dv = harness.getDataView();
    const input = ComprehensiveMessage.TestingMessage.fromBytes(dv)!;

    harness.softAssert(input.b == example.b, "byte");
    harness.softAssert(input.tf == example.tf, "bool");
    harness.softAssert(input.i16 == example.i16, "i16");
    harness.softAssert(input.ui16 == example.ui16, "ui16");
    harness.softAssert(input.i32 == example.i32, "i32");
    harness.softAssert(input.ui32 == example.ui32, "ui32");
    harness.softAssert(input.i64 == example.i64, "i64");
    harness.softAssert(input.ui64 == example.ui64, "ui64");
    harness.softAssert(input.f == Math.fround(example.f), "float");
    harness.softAssert(input.d == example.d, "double");
    harness.softAssert(input.ee == example.ee, "enumerated");
    harness.softAssert(input.es == example.es, "specified");
    harness.softAssert(input.s == example.s, "string");
    harness.softAssert(input.v2.x == Math.fround(example.v2.x), "Vec2.x");
    harness.softAssert(input.v2.y == Math.fround(example.v2.y), "Vec2.y");
    harness.softAssert(input.v3.x == Math.fround(example.v3.x), "Vec3.x");
    harness.softAssert(input.v3.y == Math.fround(example.v3.y), "Vec3.y");
    harness.softAssert(input.v3.z == Math.fround(example.v3.z), "Vec3.z");
    harness.softAssert(input.c.r == Math.fround(example.c.r), "Color.r");
    harness.softAssert(input.c.g == Math.fround(example.c.g), "Color.g");
    harness.softAssert(input.c.b == Math.fround(example.c.b), "Color.b");
    harness.softAssert(input.il.length == example.il.length, "[int16].length");
    for (let i = 0; i < input.il.length; i++) {
        harness.softAssert(input.il[i] == example.il[i], "[int16]");
    }
    harness.softAssert(input.el.length == example.el.length, "[Specified].length");
    for (let i = 0; i < input.el.length; i++) {
        harness.softAssert(input.el[i] == example.el[i], "[Specified]");
    }
    harness.softAssert(input.sl.length == example.sl.length, "[string].length");
    for (let i = 0; i < input.sl.length; i++) {
        harness.softAssert(input.sl[i] == example.sl[i], "[string]");
    }
    harness.softAssert(input.v2l.length == example.v2l.length, "[Vec2].length");
    for (let i = 0; i < input.v2l.length; i++) {
        harness.softAssert(input.v2l[i].x == example.v2l[i].x, "[Vec2].x");
        harness.softAssert(input.v2l[i].y == example.v2l[i].y, "[Vec2].y");
    }
    harness.softAssert(input.v3l.length == example.v3l.length, "[Vec3].length");
    for (let i = 0; i < input.v3l.length; i++) {
        harness.softAssert(input.v3l[i].x == example.v3l[i].x, "[Vec3].x");
        harness.softAssert(input.v3l[i].y == example.v3l[i].y, "[Vec3].y");
        harness.softAssert(input.v3l[i].z == example.v3l[i].z, "[Vec3].z");
    }
    harness.softAssert(input.cl.length == example.cl.length, "[Color].length");
    for (let i = 0; i < input.cl.length; i++) {
        harness.softAssert(input.cl[i].r == example.cl[i].r, "[Color].r");
        harness.softAssert(input.cl[i].g == example.cl[i].g, "[Color].g");
        harness.softAssert(input.cl[i].b == example.cl[i].b, "[Color].b");
    }
    harness.softAssert(input.cx.identifier == example.cx.identifier, "ComplexData.identifier");
    harness.softAssert(input.cx.label == example.cx.label, "ComplexData.label");
    harness.softAssert(input.cx.backgroundColor.r == Math.fround(example.cx.backgroundColor.r), "ComplexData.backgroundColor.r");
    harness.softAssert(input.cx.backgroundColor.g == Math.fround(example.cx.backgroundColor.g), "ComplexData.backgroundColor.g");
    harness.softAssert(input.cx.backgroundColor.b == Math.fround(example.cx.backgroundColor.b), "ComplexData.backgroundColor.b");
    harness.softAssert(input.cx.textColor.r == Math.fround(example.cx.textColor.r), "ComplexData.textColor.r");
    harness.softAssert(input.cx.textColor.g == Math.fround(example.cx.textColor.g), "ComplexData.textColor.g");
    harness.softAssert(input.cx.textColor.b == Math.fround(example.cx.textColor.b), "ComplexData.textColor.b");
    harness.softAssert(input.cx.spectrum.length == example.cx.spectrum.length, "ComplexData.spectrum.length");
    for (let i = 0; i < input.cx.spectrum.length; i++) {
        harness.softAssert(input.cx.spectrum[i].r == Math.fround(example.cx.spectrum[i].r), "ComplexData.spectrum.r");
        harness.softAssert(input.cx.spectrum[i].g == Math.fround(example.cx.spectrum[i].g), "ComplexData.spectrum.g");
        harness.softAssert(input.cx.spectrum[i].b == Math.fround(example.cx.spectrum[i].b), "ComplexData.spectrum.b");
    }
    harness.softAssert(input.cx.ranges.length == example.cx.ranges.length, "ComplexData.ranges.length");
    for (let i = 0; i < input.cx.ranges.length; i++) {
        harness.softAssert(input.cx.ranges[i] == Math.fround(example.cx.ranges[i]), "ComplexData.spectrum");
    }
    harness.softAssert(input.cxl.length == example.cxl.length, "[ComplexData].length");
    for (let i=0; i < input.cxl.length; i++) {
        harness.softAssert(input.cxl[i].identifier == example.cxl[i].identifier, "[ComplexData].identifier");
        harness.softAssert(input.cxl[i].label == example.cxl[i].label, "[ComplexData].label");
        harness.softAssert(input.cxl[i].backgroundColor.r == example.cxl[i].backgroundColor.r, "[ComplexData].backgroundColor.r");
        harness.softAssert(input.cxl[i].backgroundColor.g == example.cxl[i].backgroundColor.g, "[ComplexData].backgroundColor.g");
        harness.softAssert(input.cxl[i].backgroundColor.b == example.cxl[i].backgroundColor.b, "[ComplexData].backgroundColor.b");
        harness.softAssert(input.cxl[i].textColor.r == example.cxl[i].textColor.r, "[ComplexData].textColor.r");
        harness.softAssert(input.cxl[i].textColor.g == example.cxl[i].textColor.g, "[ComplexData].textColor.g");
        harness.softAssert(input.cxl[i].textColor.b == example.cxl[i].textColor.b, "[ComplexData].textColor.b");
        harness.softAssert(input.cxl[i].spectrum.length == example.cxl[i].spectrum.length, "[ComplexData].spectrum.length");
        for (let j = 0; j < input.cxl[i].spectrum.length; j++) {
            harness.softAssert(input.cxl[i].spectrum[j].r == example.cxl[i].spectrum[j].r, "[ComplexData].spectrum.r");
            harness.softAssert(input.cxl[i].spectrum[j].g == example.cxl[i].spectrum[j].g, "[ComplexData].spectrum.g");
            harness.softAssert(input.cxl[i].spectrum[j].b == example.cxl[i].spectrum[j].b, "[ComplexData].spectrum.b");
        }
        harness.softAssert(input.cxl[i].ranges.length == example.cxl[i].ranges.length, "[ComplexData].ranges.length");
        for (let j = 0; j < input.cxl[i].ranges.length; j++) {
            harness.softAssert(input.cxl[i].ranges[j] == example.cxl[i].ranges[j], "[ComplexData].ranges");
        }
    }
}
